presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java

```diff
@@ -0,0 +1,154 @@
+import com.facebook.presto.spi.type.TypeSignature;
+import com.google.common.collect.ImmutableList;
+import com.google.common.primitives.Ints;
+import io.airlift.slice.Slice;
+
+import java.lang.invoke.MethodHandle;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.metadata.Signature.comparableTypeParameter;
+import static com.facebook.presto.type.TypeUtils.readStructuralBlock;
+import static com.facebook.presto.type.TypeUtils.buildStructuralSlice;
+import static com.facebook.presto.type.TypeUtils.parameterizedTypeName;
+import static com.facebook.presto.util.Reflection.methodHandle;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.String.format;
+
+public final class ArrayIntersectFunction
+        extends ParametricScalar
+{
+    public static final ArrayIntersectFunction ARRAY_INTERSECT_FUNCTION = new ArrayIntersectFunction();
+    private static final String FUNCTION_NAME = "array_intersect";
+    private static final Signature SIGNATURE = new Signature(FUNCTION_NAME, ImmutableList.of(comparableTypeParameter("E")), "array<E>", ImmutableList.of("array<E>", "array<E>"), false, false);
+    private static final MethodHandle METHOD_HANDLE = methodHandle(ArrayIntersectFunction.class, "intersect", Type.class, Slice.class, Slice.class);
+
+    @Override
+    public Signature getSignature()
+    {
+        return SIGNATURE;
+    }
+
+    @Override
+    public boolean isHidden()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return true;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "Intersects elements of the two given arrays.";
+    }
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        checkArgument(types.size() == 1, format("%s expects only one argument", FUNCTION_NAME));
+        TypeSignature typeSignature = parameterizedTypeName("array", types.get("E").getTypeSignature());
+        MethodHandle methodHandle = METHOD_HANDLE.bindTo(types.get("E"));
+        Signature signature = new Signature(FUNCTION_NAME, typeSignature, typeSignature, typeSignature);
+        return new FunctionInfo(signature, getDescription(), isHidden(), methodHandle, isDeterministic(), false, ImmutableList.of(false, false));
+    }
+
+    public static Slice intersect(Type type, Slice leftArray, Slice rightArray)
+    {
+        Block leftBlock = readStructuralBlock(leftArray);
+        Block rightBlock = readStructuralBlock(rightArray);
+
+        int leftPositionCount = leftBlock.getPositionCount();
+        int rightPositionCount = rightBlock.getPositionCount();
+        int leftPosition = 0;
+        int rightPosition = 0;
+
+        List<Integer> leftPositions = Ints.asList(new int[leftPositionCount]);
```

Comment:
Use `int[]` instead to avoid allocating individual objects for each index. You'll then need to use one of the fastutils classes for the sort, but that looks pretty simple